benfulcher/hctsa
# Basic Unit Testing + Silent Install

## High Level Summary
This PR makes the following changes:
- Added a Git Actions workflow called `run-unit-tests.yaml` to execute basic unit tests when pushing to the main branch.
- Added a folder called `Tests` to the root directory which contains all files necessary for unit testing. Including:
  - `BasicPipelineTests.m`: A script which contains all basic unit tests. 
  - `HCTSA_Bonn_EEG_OG.mat`: Pre-computed HCTSA mat file for plotting/analysis functions unit tests. 
  - `INP_unit_test.mat`: File containing test time series (noisy sinusoid) for full HCTSA feature set unit test.
  - `HCTSA_catch22_expected.mat`: Pre-computed HCTSA mat file for comparing feature outputs on the catch22 feature set (w/ deterministic outputs). 
  - Modified the `install.m` script to provide a silent install option so that HCTSA can be installed on the git actions runner in non-interactive mode (i.e., does not require user input to step through installation). 

## Detailed Info
1. __Unit testing setup__:
Key steps:
1. Checkout the latest from the GitHub repository
2. Setup MATLAB (latest version) on an Ubuntu runner (latest) with all prerequisite toolboxes, cache if available.
4. Silent install HCTSA
5. Build and install TISEAN
6. Run unit tests contained in `BasicPipelineTests.m`
7. Upload the test logs as an artefact called `logfiles` which can be downloaded post-unit test for inspection.

Regarding step 7, the two test log files:
- `failed_operations_benchmark.txt`: stores all operations which returned a fatal error (quality label == 1) on the noisy sinusoid test time series.
- `summary_log.txt`: contains time benchmarks for noisy sinusoids of varying lengths and the number of operations which returned one of the eight possible quality labels (e.g., NaNs, successful, imaginary values, etc.) for the noisy sinusoid benchmark. 
These results are also printed in the Actions/Jobs section. 

2. __Silent install__:
`install.m` will now accept a `silentInstall` flag which defaults to `false` if no user input is provided. If install is called with the `silentInstall` flag as `install(true)`, HCTSA will be installed without asking for user input. This will allow HCTSA to be installed on the Git Actions runner and other non-interactive environments. 

 __The Unit Tests__
The unit tests are quite basic and their intended purpose is to run through the typical HCTSA pipeline from computing features to plotting/analysing the results and flag any fatal errors (i.e., does the function break?). These unit tests will not distinguish between different errors/exceptions. In short, if an error is caught, the test will fail the assertion check. 

__Coverage__:
The unit tests cover basic testing of most of the plotting/analysis functions as well as computation of the full HCTSA feature set + catch22 feature set on a test time series (noisy sinusoid). Computation time benchmarks are performed for three sinusoids of varying length (500, 1000, 5000 samples).
